compiler/rustc_type_ir/src/elaborate.rs

```diff
@@ -89,6 +90,55 @@ impl<I: Interner, O: Elaboratable<I>> Elaborator<I, O> {
                     return;
                 }
 
+                // N.B. this is a bit of a hack to get implied bounds for effects working..
+                if cx.is_lang_item(data.def_id(), TraitSolverLangItem::EffectsCompat)
+                    && matches!(self.mode, Filter::All)
+                {
+                    // first, ensure that the predicate we've got looks like a `<T as Tr>::Fx: EffectsCompat<somebool>`.
+                    // our plan is to gather information such that we elaborate into `<T as SuperTr>::Fx: EffectsCompat<somebool>`
+                    // when we know that `trait Tr: ~const SuperTr`.
+                    if let ty::Alias(ty::AliasTyKind::Projection, alias_ty) = data.self_ty().kind()
+                    {
+                        // look for effects-level bounds that look like `<Self as Tr>::Fx: TyCompat<<Self as SuperTr>::Fx>`,
+                        // which is proof to us that `Tr: ~const SuperTr`.
+                        let bounds = cx.explicit_implied_predicates_of(cx.parent(alias_ty.def_id));
+                        let expected_args_with_self =
+                            I::GenericArgs::identity_for_item(cx, alias_ty.def_id);
+                        let elaborated = bounds.iter_identity().filter_map(|(clause, _)| {
```

Comment:
Could you explain with more detail exactly what this is doing? There's a lot of `data`/`data2`/`ty`/`ty2`/etc that I have no idea what they signify.

Reply:
I added more comments to explain the process and also made the process many times easier to understand (I'd even say the previous version was pretty wrong)